maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java

```diff
@@ -368,71 +351,113 @@ private void restoreCommandLineProperties( ArchetypeConfiguration archetypeConfi
         {
-            if ( "groupId".equals( left ) )
-            {
-                return -1;
-            }
-            if ( "groupId".equals( right ) )
-            {
-                return 1;
-            }
-            if ( "artifactId".equals( left ) )
-            {
-                return -1;
-            }
-            if ( "artifactId".equals( right ) )
+            Map<String, Set<String>> result = new HashMap<>();
+
+            List<String> requiredProperties = archetypeConfiguration.getRequiredProperties();
+
+            for ( String propertyName : requiredProperties )
             {
-                return 1;
+                final Set<String> referencedPropertyNames = new LinkedHashSet<>();
+
+                String defaultValue = archetypeConfiguration.getDefaultValue( propertyName );
+                if ( StringUtils.contains( defaultValue, "${" ) )
+                {
+                    try
+                    {
+                        final boolean dumpNamespace = false;
+                        SimpleNode node = RuntimeSingleton.parse(
+                                        new StringReader( defaultValue ), propertyName + ".default", dumpNamespace );
+                        node.jjtAccept( new BaseVisitor()
+                        {
+                            @SuppressWarnings( "unchecked" )
+                            @Override
+                            public Object visit( ASTReference node, Object data )
+                            {
+                                ( ( Set<String> ) data ).add( node.getFirstToken().next.image );
+                                return super.visit( node, data );
+                            }
+                        }, referencedPropertyNames );
+                    }
+                    catch ( ParseException e )
+                    {
+                    }
+                }
+
+                result.put( propertyName, referencedPropertyNames );
             }
-            if ( "version".equals( left ) )
+
+            return result;
+        }
+
+        /**
+         * Yoda semantics.
+         * @param targetProperty
+         * @param sourceProperty
+         * @return {@code boolean}
+         */
+        private boolean references( String targetProperty, String sourceProperty )
+        {
+            if ( targetProperty.equals( sourceProperty ) )
             {
-                return -1;
+                return false;
             }
-            if ( "version".equals( right ) )
+            synchronized ( this )
             {
-                return 1;
+                if ( ! propertyReferenceMap.containsKey( sourceProperty ) )
+                // something has changed
+                {
+                   this.propertyReferenceMap = computePropertyReferences(); 
+                }
             }
-            if ( "package".equals( left ) )
+            Set<String> references = propertyReferenceMap.get( sourceProperty );
```

Comment:
It's mildly confusing to have a local variable and method with the same name. Can you rename one?